File: src/test/java/com/dozycoffee/wms/warehouse/fixture/WorkAreaTestBuilder.java
package com.dozycoffee.wms.warehouse.fixture;

import com.dozycoffee.wms.warehouse.domain.enumeration.AreaCode;
import com.dozycoffee.wms.warehouse.domain.enumeration.WorkAreaStatus;
import com.dozycoffee.wms.warehouse.domain.model.WorkArea;

public class WorkAreaTestBuilder {

    private Long workAreaId = null;
    private Long warehouseId = 1L;
    private AreaCode areaCode = AreaCode.INBOUND;
    private int usedCapacity = 0;
    private WorkAreaStatus workAreaStatus = WorkAreaStatus.ACTIVE;

    public static WorkAreaTestBuilder workArea() {
        return new WorkAreaTestBuilder();
    }

    public WorkAreaTestBuilder workAreaId(Long workAreaId) {
        this.workAreaId = workAreaId;
        return this;
    }

    public WorkAreaTestBuilder warehouseId(Long warehouseId) {
        this.warehouseId = warehouseId;
        return this;
    }

    public WorkAreaTestBuilder areaCode(AreaCode areaCode) {
        this.areaCode = areaCode;
        return this;
    }

    public WorkAreaTestBuilder usedCapacity(int usedCapacity) {
        this.usedCapacity = usedCapacity;
        return this;
    }

    public WorkAreaTestBuilder workAreaStatus(WorkAreaStatus workAreaStatus) {
        this.workAreaStatus = workAreaStatus;
        return this;
    }

    public WorkArea build() {
        if (workAreaId != null) {
            return WorkArea.reconstitute(
                    workAreaId,
                    warehouseId,
                    areaCode,
                    usedCapacity,
                    workAreaStatus
            );
        }
        return WorkArea.create(
                warehouseId,
                areaCode,
                workAreaStatus
        );
    }
}

File: src/test/kotlin/com/dozycoffee/wms/warehouse/fixture/WorkAreaTestBuilder.kt
package com.dozycoffee.wms.warehouse.fixture

import com.dozycoffee.wms.warehouse.domain.enumeration.AreaCode
import com.dozycoffee.wms.warehouse.domain.enumeration.AvailabilityStatus
import com.dozycoffee.wms.warehouse.domain.model.WorkArea

class WorkAreaTestBuilder {

    private var workAreaId: Long? = null
    private var warehouseId: Long? = 1L
    private var areaCode: AreaCode? = AreaCode.INBOUND
    private var usedCapacity: Int = 0
    private var workAreaStatus: AvailabilityStatus? = AvailabilityStatus.AVAILABLE

    companion object {
        fun workArea(): WorkAreaTestBuilder = WorkAreaTestBuilder()
    }

    fun workAreaId(workAreaId: Long?): WorkAreaTestBuilder {
        this.workAreaId = workAreaId
        return this
    }

    fun warehouseId(warehouseId: Long?): WorkAreaTestBuilder {
        this.warehouseId = warehouseId
        return this
    }

    fun areaCode(areaCode: AreaCode?): WorkAreaTestBuilder {
        this.areaCode = areaCode
        return this
    }

    fun usedCapacity(usedCapacity: Int): WorkAreaTestBuilder {
        this.usedCapacity = usedCapacity
        return this
    }

    fun workAreaStatus(workAreaStatus: AvailabilityStatus?): WorkAreaTestBuilder {
        this.workAreaStatus = workAreaStatus
        return this
    }

    fun build(): WorkArea {
        val id = workAreaId
        if (id != null) {
            return WorkArea.reconstitute(
                id,
                requireNotNull(warehouseId),
                requireNotNull(areaCode),
                usedCapacity,
                requireNotNull(workAreaStatus)
            )
        }
        return WorkArea.create(warehouseId, areaCode, workAreaStatus)
    }
}
